gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java

```diff
@@ -0,0 +1,189 @@
+
+  /**
+   * Listeners of JobCatalog, Like GobblinInstaceDriver.
+   * Distinguished it from JobSpec Listeners.
+   */
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(FSJobCatalog.class);
+
+  // A monitor for changes to job conf files for general FS
+  // This embedded monitor is monitoring job configuration files instead of JobSpec Object.
+  private final PathAlterationDetector pathAlterationDetector;
+
+  private final PathAlterationObserver observer;
+
+  /**
+   * Initialize the JobCatalog, fetch all jobs in jobConfDirPath.
+   * @param sysConfig
+   * @throws Exception
+   */
+  public FSJobCatalog(Config sysConfig)
+      throws Exception {
+    this(sysConfig, null);
+    /* Automatically shceudle the checkAndNotify method using detector's threadpool, in general usage. */
+    this.pathAlterationDetector.start();
+  }
+
+  /**
+   * The expose of observer is used for testing purpose, so that
+   * the checkAndNotify method can be revoked manually, instead of waiting for
+   * the scheduling timing.
+   * @param sysConfig The same as general constructor.
+   * @param observer The user-initialized observer.
+   * @throws Exception
+   */
+  @VisibleForTesting
+  protected FSJobCatalog(Config sysConfig, PathAlterationObserver observer)
+      throws Exception {
+    super(sysConfig);
+    Properties sysProp = ConfigUtils.configToProperties(this.sysConfig);
+    this.observer = observer;
+
+    long pollingInterval = Long.parseLong(
+        sysProp.getProperty(ConfigurationKeys.JOB_CONFIG_FILE_MONITOR_POLLING_INTERVAL_KEY,
+            Long.toString(ConfigurationKeys.DEFAULT_JOB_CONFIG_FILE_MONITOR_POLLING_INTERVAL)));
+    this.pathAlterationDetector = new PathAlterationDetector(pollingInterval);
+    Optional<PathAlterationObserver> observerOptional = Optional.fromNullable(this.observer);
+    FSPathAlterationListenerAdaptor configFilelistener = new FSPathAlterationListenerAdaptor(this.jobConfDirPath);
+    FSJobCatalogHelper.addPathAlterationObserver(this.pathAlterationDetector, configFilelistener, observerOptional,
+        this.jobConfDirPath);
+  }
+
+  /**
+   * Allow user to programmatically add a new JobSpec.
+   * The method will materialized the jobSpec into real file.
+   *
+   * @param jobSpec The target JobSpec Object to be materialized.
+   *                Noted that the URI return by getUri is a relative path.
+   */
+  @Override
+  public synchronized void put(JobSpec jobSpec) {
+    Preconditions.checkNotNull(jobSpec);
+    try {
+      String srcFileName = jobSpec.getUri().toString();
+      Path tmpPath = PathUtils.mergePaths(this.jobConfDirPath, new Path(jobSpec.getUri().toString()));
+      if (fs.exists(tmpPath)) {
+        LOGGER.info("The job with URI[" + tmpPath + "] has been added before, will cover the original one.");
+
+        FSJobCatalogHelper.materializedJobSpec(new Path(srcFileName), this.jobConfDirPath, jobSpec, true);
```

Comment:
The last argument (`true`) should be a call to a method `e.g. shouldLoadGlobalConf()`, so that the mutable subclass can override this to `false`.